ui - enhance flow typing

JBWatenbergScality opened the pull request:
**Component**:

ui, typing

**Context**: 

This pull request is a first milestone to introduce better typings in the project. 
This brings type-safety and enhanced developer experience to the project.

As introducing strong typing everywhere in the project would require a few more pull requests the idea of this one is to focus on kubernetes client and reducers that make use of kubernetes client. However it also brings typings for few others reducers and introduce starting points to expand typing to other areas in the project.


Closes: #2999

See: #3000 #3001

JBWatenbergScality committed:
ui: Add kubernetes client typing to client wrapper
Refs: https://github.com/scality/metalk8s/issues/2999

JBWatenbergScality committed:
ui: Add volume reducer typing
Refs: https://github.com/scality/metalk8s/issues/2999

JBWatenbergScality committed:
ui: Add constant typing
Refs: https://github.com/scality/metalk8s/issues/2999

JBWatenbergScality committed:
ui: Add config and theme typing
Refs: https://github.com/scality/metalk8s/issues/2999

JBWatenbergScality committed:
ui: Start monitoring typing
Refs: https://github.com/scality/metalk8s/issues/3001

JBWatenbergScality committed:
ui: Start node typing
Refs: https://github.com/scality/metalk8s/issues/3000

JBWatenbergScality committed:
ui: Add notification typings

JBWatenbergScality committed:
ui: Add pods reducer typing
Refs: https://github.com/scality/metalk8s/issues/2999

JBWatenbergScality committed:
ui: Start salt reducer typing
Refs: https://github.com/scality/metalk8s/issues/3000

JBWatenbergScality committed:
ui: Start solutions reducer typing
Refs: https://github.com/scality/metalk8s/issues/2999

JBWatenbergScality committed:
ui: Add prometheus alert type

JBWatenbergScality committed:
ui: Add root state typing

JBWatenbergScality committed:
ui: Introduce useTypedSelector in volume page

bert-e commented:
# Hello jbwatenbergscality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/flow-typing` has diverged from
`development/2.7` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.7` into `improvement/flow-typing`
* Rebase `improvement/flow-typing` onto `origin/development/2.7`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

gdemonet commented on ui/src/services/k8s/api.js:
```diff
@@ -1,16 +1,60 @@
+  kind: 'Volume',
+  metadata?: V1ObjectMeta,
+  spec: {
+    mode: 'Filesystem',
```
The mode can be either `Filesystem` or `Block`, actually. See [our docs](https://metal-k8s.readthedocs.io/en/latest/operation/volume_management/volume_creation_deletion_cli.html#volume-creation) or [the CRD](https://github.com/scality/metalk8s/blob/development/2.7/storage-operator/deploy/crds/storage.metalk8s.scality.com_volumes_crd.yaml#L43-L49)

gdemonet commented on ui/src/services/k8s/api.js:
```diff
@@ -1,16 +1,60 @@
+    storageClassName: string
+  },
+  status: {
+    deviceName: string,
```
If this was not yet reconciled (e.g. because the Volume is `Pending`), this will be empty from the Operator's PoV, and absent in the API object.

gdemonet commented on ui/src/services/k8s/api.js:
```diff
@@ -1,16 +1,60 @@
+    },
+    storageClassName: string
+  },
+  status: {
```
If you want to be exhaustive, there is also a `status.job` key (which is also optional, and its usage is internal to the Operator).

gdemonet commented on ui/src/ducks/app/volumes.js:
```diff
@@ -138,42 +156,48 @@ export const refreshPersistentVolumesAction = () => {
 };
 
-export const fetchCurrentVolumeObjectAction = (volumeName) => {
+export const fetchCurrentVolumeObjectAction = (volumeName: any) => {
```
I guess we can safely assume this should be a `string`, no?

gdemonet commented on ui/src/ducks/app/volumes.js:
```diff
@@ -138,42 +156,48 @@ export const refreshPersistentVolumesAction = () => {
 
 // Sagas
-export function* fetchVolumes() {
+export function* fetchVolumes(): Generator<{
+  body: {items: Metalk8sV1Alpha1Volume[]};
+} | {error: any, body: null}, {
+  body: {items: Metalk8sV1Alpha1Volume[]};
+} | {error: any, body: null}, {
+  body: {items: Metalk8sV1Alpha1Volume[]};
+} | {error: any, body: null}> {
```
I'm not sure that's a valid type for sagas. IIUC, the type signature for a coroutine is `Generator<Yield, Return, Next>`. Sagas only yield "effects" (in the sense of `redux-saga`), and they may be fed with `any`thing.
For this reason, there is a handy type definition in `flow-typed` for the `redux-saga` library:

```javascript
import type { Saga } from 'redux-saga';

// same as:
declare type Saga<T> = Generator<Effect, T, any>
```

See the definitions [here](https://github.com/flow-typed/flow-typed/blob/master/definitions/npm/redux-saga_v1.x.x/flow_v0.104.x-/redux-saga_v1.x.x.js)

JBWatenbergScality replied:
Thanks I was looking for those typings desperately 👍  👍

gdemonet commented on ui/src/services/api.js:
```diff
@@ -1,15 +1,69 @@
 }
 
-export function fetchTheme() {
+export type Brand = {
```
OK for now, but IMHO this type (`Theme` as well) should live in `@scality/core-ui`.

gdemonet commented on ui/src/services/api.js:
```diff
@@ -1,15 +1,69 @@
+export type WrappedThemes = {
+  "theme": Themes,
+  "default": "dark"
+}
```
I'm not entirely sure if this is the intended type (maybe it's what the code expects today, didn't check). For me, the goal is more to have something like:

```javascript
export type WrappedThemes = {
  "theme": { [key: string]: Theme },
  "default": string
}
```

gdemonet commented on ui/src/ducks/app/nodes.js:
```diff
@@ -161,6 +162,25 @@ const defaultState = {
+  isRefreshing: boolean,
+  isLoading: boolean,
+  IPsInfo: any,
+  currentNodeObject: any,
```
Missing some `TODO` comments?

gdemonet commented on ui/src/ducks/app/pods.js:
```diff
@@ -44,7 +64,7 @@ export function* fetchPods() {
               ? pod.status.containerStatuses[0].restartCount
               : 0,
-          volumes: pod?.spec?.volumes?.map((volume) => ({
+          volumes: (pod?.spec?.volumes || []).map((volume) => ({
```
Shouldn't we use the nullish coallescing operator for that?

```javascript
pod?.spec?.volumes ?? []
```

gdemonet commented on ui/src/ducks/app/solutions.js:
```diff
@@ -130,10 +137,10 @@ export function* createEnvironment(action) {
   const { envName, solName, solVersion } = action.payload;
 
-  const existingEnv = yield select((state) => state.app.solutions.environments);
+  const existingEnv: SolutionsApi.Environments = yield select((state) => state.app.solutions.environments);
```
I think that's a better approach to type effect returns (especially for `call`s) than trying to type the coroutine itself. May be worth generalizing to other sagas.

gdemonet replied:
I'm a bit surprised though, don't we have a type mismatch? The annotation L.49 gives:

```javascript
  environments: SolutionsApi.Environment[],
```

But the type definition `SolutionsApi.Environments` describes a map (`{ [name: string]: Environment }`), not a list.

gdemonet commented on ui/src/services/alertmanager/api.js:
```diff
@@ -1,11 +1,35 @@
 import ApiClient from '../ApiClient';
 
-let alertmanagerApiClient = null;
+let alertmanagerApiClient: ApiClient = null;
```
Shouldn't it be a `Maybe` type?

```javascript
let alertmanagerApiClient: ?ApiClient = null;
```

gdemonet commented on ui/src/services/alertmanager/api.js:
```diff
@@ -1,11 +1,35 @@
 
-export function getAlerts() {
+export type PrometheusAlert = {
+  "annotations": {
+    [key: string]: string,
+  },
```
OK for now, but we'll need to type some of the expected keys (e.g. `description`, `message`, ...). We've been hit by a missing key recently (see #2994).

JBWatenbergScality replied:
if they are always present in the alerts I can add typing for `description` and `summary`. I'm just unsure Prometheus alert manager will always send those keys as the documentation is not mentioning this.

gdemonet replied:
No they are optional indeed, so not sure if Flow can help in any way here.

gdemonet commented on ui/src/ducks/app/alerts.js:
```diff
@@ -35,7 +42,7 @@ export const fetchAlertsAlertmanagerAction = () => {
 };
 
-export const updateAlertsAlertmanagerAction = (payload) => {
+export const updateAlertsAlertmanagerAction = (payload: $Shape<AlertsState>) => {
```
Oh, neat type 😄 , I guess we'll be able to use it for many "update" action creators.

gdemonet commented on ui/src/ducks/config.js:
```diff
@@ -56,7 +59,26 @@ const defaultState = {
+export type ConfigState = {
+  language: string,
+  theme: Theme,
+  api: Config | null,
```
`?Config` would work here as well, IIUC

gdemonet commented on ui/src/ducks/config.js:
```diff
@@ -100,48 +122,50 @@ export function fetchConfigAction() {
 // Selectors
-export const languageSelector = (state) => state.config.language;
-export const apiConfigSelector = (state) => state.config.api;
+export const languageSelector = (state: RootState) => state.config.language;
```
Just a question: since `RootState` is properly typed, is flow able to infer the return type of these selectors?

JBWatenbergScality replied:
Yes, flow is inferring the type of `languageSelector` from `state.config.language`

gdemonet commented on ui/src/ducks/login.js:
```diff
@@ -28,23 +36,19 @@ export default function reducer(state = defaultState, action = {}) {
-  const result = yield call(ApiSalt.authenticate, user);
-  if (!result.error) {
+export function* authenticateSaltApi(): Generator<any, void, any> {
+  const api: Config | null = yield select(apiConfigSelector);
```
(ugh, even if flow was able to infer the return type of a selector, it can't guess the value injected by the coroutine's "owner"... some big limitation of sagas here)

JBWatenbergScality replied:
Yes the simplest way I found to type complex saga with different return type over iterations is to type `Next` as any and  force the type of each of the variables 😞 . Even through this is not directly linked to redux but more to typing, IMO redux and redux saga are adding a lot of complexity for such simple tasks as data fetching, I prefer using libraries such as  `react-query` or `swr` that greatly ease data fetching and caching use cases and that are way more easier to type as in most cases the typing would be inferred from API clients typing.

gdemonet commented on ui/src/ducks/login.js:
```diff
@@ -28,23 +36,19 @@ export default function reducer(state = defaultState, action = {}) {
+export function* authenticateSaltApi(): Generator<any, void, any> {
+  const api: Config | null = yield select(apiConfigSelector);
+  const user: User = yield select((state: RootState) => state.oidc?.user);
+  const result: { error: any } | ApiSalt.SaltToken = yield call(ApiSalt.authenticate, user);
```
Could be handy to have type aliases for such "result" values:

```javascript
declare type Result<T> = { error: any } | T

// Or for API results under a `body` key
declare type APIResult<T> = Result<{ body: T }>
```

Here we'd have:

```javascript
const result: Result<Api.SaltToken> = yield call(...)
```

gdemonet commented on ui/src/ducks/reducer.js:
```diff
@@ -29,4 +38,26 @@ const rootReducer = combineReducers({
+  }
+}
+
+export const useTypedSelector: <TSelected>(
+  selector: (state: RootState) => TSelected,
+  equalityFn?: (left: TSelected, right: TSelected) => boolean
+) => TSelected = useSelector;
```
Two comments:

- I think we have defined custom hooks in `src/services/utils.js`(though granted, it doesn't make much sense either) - IMHO, adding a plain `src/hooks.js` could make it easier to find in the future
- the purpose of this custom hook is not obvious at first glance, so I'd suggest we add a comment to explain it (and remind the reader that they should use it instead of the standard `useSelector` hook)

gdemonet reviewed (changes requested):
Super cool stuff, thanks for starting this heavy lifting, I'm confident this will increase productivity a lot!
Some comments and suggestions here and there 😉 , just adding a more general one here: are you running `Prettier`? We have a committed `.prettierrc` for you to use, the goal being to keep the code consistent without spending time on aesthetics in PR reviews 😜

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/flow-typing` and the
destination branch `development/2.7`.

Please resolve the conflict on **the feature branch** (`improvement/flow-typing`).

```bash
 $ git fetch
 $ git checkout origin/improvement/flow-typing
 $ git merge origin/development/2.7
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/flow-typing
```

JBWatenbergScality committed:
ui: Add a flow typed client generator

The goal of this commit is to introduce a generator of a client based upon `@kubernetes/client` custom objects client from kubernetes `CustomResourceDefinition`.

The idea behind this is to have always up to date types definition for our custom resource definition and be in sync with the json schema contained within those definitions.

Refs: #2999

JBWatenbergScality committed:
ui: Use redux-saga Effect in saga typings

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

JBWatenbergScality committed:
ui: Prettier all the things

gdemonet commented on ui-metalk8s-client-generator/README.md:
```diff
@@ -0,0 +1,23 @@
+## Metalk8s client generator
+
+It generates a flow client based on `@kubernetes/client` custom objects client from kubernetes `CustomResourceDefinition` of metalk8s volumes.
```
Awesome :+1:

JBWatenbergScality committed:
ui: Add Namespaced scope support to meatlk8s client generator

ChengYanJin reviewed (dismissed):
LGTM! 
Can't wait to get it to merge and benefit from the typing.

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/flow-typing` and the
destination branch `development/2.7`.

Please resolve the conflict on **the feature branch** (`improvement/flow-typing`).

```bash
 $ git fetch
 $ git checkout origin/improvement/flow-typing
 $ git merge origin/development/2.7
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/flow-typing
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/flow-typing` into
`development/2.7`
* [w/2.8/improvement/flow-typing](https://github.com/scality/metalk8s/compare/development/2.8...w/2.8/improvement/flow-typing) will be merged into `development/2.8` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

gdemonet reviewed (changes requested):
Could we squash the commits into logical units, now that reviewers approved? I'm not too fond of `fix stuff` commits, especially for fixing things introduced in the same PR.

JBWatenbergScality committed:
ui: Introduce useTypedSelector to enable simpler typed selection of redux store data

JBWatenbergScality committed:
ui: Introduce result type to represent an entity or an error

bert-e commented:
# History mismatch



Merge commit #6e03a09581f9c41616f1bf5816c34e72ea53663c on the integration branch
`w/2.8/improvement/flow-typing` is merging a branch which is neither the current
branch `improvement/flow-typing` nor the development branch
`development/2.8`.

It is likely due to a rebase of the branch `improvement/flow-typing` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented on ui-metalk8s-client-generator/src/generateClient.ts:
```diff
@@ -55,6 +55,22 @@ export function generateClient(crdFile: string, destinationFile: string) {
+    } catch (error) {
+      return { error };
+    }
+  }
```
Maybe this part should be squashed with `ui: add a flow typed client generator` (it's more than just `ui: use generated metalk8s client and types`)

JBWatenbergScality committed:
ui: Use generated metalk8s client and types
Refs: https://github.com/scality/metalk8s/issues/2999

JBWatenbergScality commented:
/approve

bert-e commented:
# History mismatch



Merge commit #6e03a09581f9c41616f1bf5816c34e72ea53663c on the integration branch
`w/2.8/improvement/flow-typing` is merging a branch which is neither the current
branch `improvement/flow-typing` nor the development branch
`development/2.8`.

It is likely due to a rebase of the branch `improvement/flow-typing` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

JBWatenbergScality commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/flow-typing` into
`development/2.7`
* [w/2.8/improvement/flow-typing](https://github.com/scality/metalk8s/compare/development/2.8...w/2.8/improvement/flow-typing) will be merged into `development/2.8` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/ffbed035356c9f798de742f0e850fbad481aa582) did not succeed in branch improvement/flow-typing.




*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #16ab5607ac98a462b7293386476c98abd9508bbf on the integration branch
`w/2.8/improvement/flow-typing` is merging a branch which is neither the current
branch `improvement/flow-typing` nor the development branch
`development/2.8`.

It is likely due to a rebase of the branch `improvement/flow-typing` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

JBWatenbergScality commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/flow-typing` into
`development/2.7`
* [w/2.8/improvement/flow-typing](https://github.com/scality/metalk8s/compare/development/2.8...w/2.8/improvement/flow-typing) will be merged into `development/2.8` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality










*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/708300681fecb53b436eff529705c913d0f9ec32) did not succeed in branch improvement/flow-typing.




*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #9a1bb97024ac924946876225330345dcd94bf116 on the integration branch
`w/2.8/improvement/flow-typing` is merging a branch which is neither the current
branch `improvement/flow-typing` nor the development branch
`development/2.8`.

It is likely due to a rebase of the branch `improvement/flow-typing` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

JBWatenbergScality commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/flow-typing` into
`development/2.7`
* [w/2.8/improvement/flow-typing](https://github.com/scality/metalk8s/compare/development/2.8...w/2.8/improvement/flow-typing) will be merged into `development/2.8` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.8`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.8`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`



Please check the status of the associated issue None.

Goodbye jbwatenbergscality.
